File: Python3/1 - 200/98 Validate Binary Search Tree.py
# 52ms, 87.32%; 15.7MB, 11.45%
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        return self.__check(root, None, None)

    def __check(self, root: TreeNode, min_limit: int, max_limit: int) -> bool:
        if not root: return True
        if min_limit is not None and root.val <= min_limit: return False
        if max_limit is not None and root.val >= max_limit: return False
        return self.__check(root.left, min_limit, root.val) and self.__check(root.right, root.val, max_limit)

File: 1 - 200/98. Validate Binary Search Tree.cc
// 16ms, 99.87%; 20.7MB, 27.46%
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
public:
    bool isValidBST(TreeNode *root) {
        return check(root);
    }

private:
    bool check(TreeNode *root, long long minLimit = LONG_LONG_MIN, long long maxLimit = LONG_LONG_MAX) {
        if (!root) return true;
        if (root->val <= minLimit || root->val >= maxLimit) return false;
        return check(root->left, minLimit, root->val) && check(root->right, root->val, maxLimit);
    }
};
